File: competition/07-tree/tree.lua
-- Build a perfect binary tree, then fold it — recursive data + dispatch.
local function build(depth, start)
  if depth == 0 then return { v = start } end
  return { l = build(depth - 1, start), r = build(depth - 1, start + 1) }
end

local function tree_sum(t)
  if t.l == nil then return t.v end
  return tree_sum(t.l) + tree_sum(t.r)
end

print(tree_sum(build(21, 1)))

File: competition/07-tree/tree.hs
-- Build a perfect binary tree of nominal nodes, then fold it — recursive data
-- construction and pattern matching at depth.
data Tree = Leaf !Int | Node Tree Tree

build :: Int -> Int -> Tree
build 0 start = Leaf start
build depth start = Node (build (depth - 1) start) (build (depth - 1) (start + 1))

treeSum :: Tree -> Int
treeSum (Leaf v) = v
treeSum (Node l r) = treeSum l + treeSum r

main :: IO ()
main = print (treeSum (build 21 1))
